### PR TITLE
Linked list impl

### PR DESCRIPTION
**Pull request by IParov**

### CODE REVIEW

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 1]
```diff
@@ -3,45 +3,157 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
+    private static class Node<E> {
```
Comment (andriy-shulepa) [inline src/main/java/core/basesyntax/MyLinkedList.java:10]: move inner class to the bottom of the file

**Review by andriy-shulepa — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 2]
```diff
@@ -3,45 +3,148 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        Node<T> lastNode = last;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:12]:
```suggestion
        Node<T> oldLast = last;
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 3]
```diff
@@ -3,45 +3,148 @@
...
+        return elementByIndex.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = first;
+        for (int i = 0; i < size; i++) {
+            if ((node.value != null && node.value.equals(object)) || node.value == object) {
+                unlink(node);
+                size--;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:80]: size--; is part of unlink logic. Is it possible to unlink node without size change?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 4]
```diff
@@ -3,45 +3,148 @@
...
+        return size == 0;
+    }
+
+    private void unlink(Node<T> node) {
+        if (size == 1) {
+            first = null;
+            last = null;
+            return;
+        }
+        if (node.equals(first)) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:104]: equals... Is it possible? I do not see equals overriding in class Node
Better not to use equals in unlink()

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 5]
```diff
@@ -3,45 +3,148 @@
...
+        checkIndex(index);
+        return findElementByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = findElementByIndex(index).value;
+        findElementByIndex(index).value = value;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:61]: do not call findElementByIndex 2 times

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 6]
```diff
@@ -3,45 +3,148 @@
...
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:29]:
```suggestion
```
`checkIndex()` should be part of `findElementByIndex()`

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 7]
```diff
@@ -3,45 +3,148 @@
...
+            last = node.prev;
+            return;
+        }
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " " + "is not valid.");
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:118]: add size info

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 3 above)
Comment (IParov) [reply src/main/java/core/basesyntax/MyLinkedList.java:80]: Looks like we don't need the method "remove" at all.

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 4 above)
Comment (IParov) [reply src/main/java/core/basesyntax/MyLinkedList.java:104]: Yes, it's possible and works fine.  Ok, I will use "==" instead.

**Review by Rommelua — APPROVED** [review]

**Pull request closed without merging**